model/criteria.py

```diff
@@ -192,3 +192,110 @@ def check_criteria_against_paths(self, control_flow_graph, execution_paths):
 
     def __repr__(self):
         return 'i - TB - All {} loops'.format(self.i)
+
+
+class TDef(Criteria):
+    """
+    We get all the vertices containing an assigment from the cover graph and
+    check if they are in some of the path of the execution_paths we got from
+    execution.
+    """
+
+    def check_criteria_against_paths(self, control_flow_graph, execution_paths):
+        prog_vars = control_flow_graph.get_variables()
+        # we get all the vertices where there is an reference in the graph
+        ref_vertices = []
+        ref_variables = set()
+        for var in prog_vars:
+            for vertex in control_flow_graph.vertices:
+                if var in control_flow_graph.get_ref_variables(vertex):
+                    ref_vertices.append((var, vertex))
+                    ref_variables.add(var)
+                    self.to_cover.append(vertex)
+
+        # we also get all the vertices where there is a definition in the graph
+        def_vertices = []
+        for var in ref_variables:
+            for vertex in control_flow_graph.vertices:
+                if var in control_flow_graph.get_def_variables(vertex):
+                    def_vertices.append((var, vertex))
+
+        # getting all vertices where variables are effectively defined during the
+        # execution
+        def_in_exec_path = []
+        for var in prog_vars:
+            for path in execution_paths:
+                for vertex in path:
+                    for var2, vertex2 in def_vertices:
+                        if var == var2 and vertex == vertex2:
+                            def_in_exec_path.append(var)
+
+        # checking that those variables are effectively used
+        for var in def_in_exec_path:
+            is_used_once = False
+            for path in execution_paths:
+                for vertex in path:
+                    for var2, vertex2 in ref_vertices:
+                        if var == var2 and vertex == vertex2:
+                            self.covered.append(vertex)
+                            is_used_once = True
+            if not is_used_once:
+                return False
+        return True
```

Comment:
I have a problem with this criteria. I try to execute the code in main.py with our fact program and I get several time the same vertex to cover.